[pytorch][quant] Add assert for min, max, qmin, qmax for ChooseQuantizationParams

jianyuh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33167 [pytorch][quant] Debug NaN issues in quantized RNN
* **#32739 [pytorch][quant] Add assert for min, max, qmin, qmax for ChooseQuantizationParams**

As Title says.

Differential Revision: [D19610810](https://our.internmc.facebook.com/intern/diff/D19610810/)

jianyuh committed:
[pytorch][quant] Add assert for min, max, qmin, qmax for ChooseQuantizationParams

As Title says.

Differential Revision: [D19610810](https://our.internmc.facebook.com/intern/diff/D19610810/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit c10ed6b9:


**None of the build failures appear to be your fault**.

* **1/1** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=c10ed6b9d6ecde05d56ac5f79b8ed48813ecd973).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4364432) caffe2_onnx_py3_6_clang7_ubuntu16_04_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92286250) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92286250)) :snowflake:
<details>
<summary>
<code>Jan 29 00:36:45 test/onnx/test_utility_funs.py::TestUtilityFuns::test_constant_fold_reshape Fatal Python error: Segmentation fault</code>
</summary>

```
Jan 29 00:36:44 test/onnx/test_utility_funs.py::TestUtilityFuns_opset11::test_constant_fold_transpose_matmul PASSED [ 99%] 
Jan 29 00:36:44 test/onnx/test_utility_funs.py::TestUtilityFuns_opset11::test_constant_fold_unsqueeze PASSED [ 99%] 
Jan 29 00:36:44 test/onnx/test_utility_funs.py::TestUtilityFuns_opset11::test_error_on_data_parallel PASSED [ 99%] 
Jan 29 00:36:44 test/onnx/test_utility_funs.py::TestUtilityFuns_opset11::test_is_in_onnx_export PASSED [ 99%] 
Jan 29 00:36:44 test/onnx/test_utility_funs.py::TestUtilityFuns_opset11::test_strip_doc_string PASSED [ 99%] 
Jan 29 00:36:44 test/onnx/test_utility_funs.py::TestUtilityFuns_opset11::test_validate_dynamic_axes_invalid_input_output_name PASSED [ 99%] 
Jan 29 00:36:44 test/onnx/test_utility_funs.py::TestUtilityFuns::test_constant_fold_concat PASSED [ 99%] 
Jan 29 00:36:44 test/onnx/test_utility_funs.py::TestUtilityFuns::test_constant_fold_div PASSED [ 99%] 
Jan 29 00:36:44 test/onnx/test_utility_funs.py::TestUtilityFuns::test_constant_fold_lstm PASSED [ 99%] 
Jan 29 00:36:44 test/onnx/test_utility_funs.py::TestUtilityFuns::test_constant_fold_mul PASSED [ 99%] 
Jan 29 00:36:45 test/onnx/test_utility_funs.py::TestUtilityFuns::test_constant_fold_reshape Fatal Python error: Segmentation fault 
Jan 29 00:36:45  
Jan 29 00:36:45 Current thread 0x00007f1b97f7d700 (most recent call first): 
Jan 29 00:36:45   File "/usr/local/lib/python3.6/dist-packages/_pytest/_io/saferepr.py", line 43 in repr_instance 
Jan 29 00:36:45   File "/usr/lib/python3.6/reprlib.py", line 65 in repr1 
Jan 29 00:36:45   File "/usr/lib/python3.6/reprlib.py", line 55 in repr 
Jan 29 00:36:45   File "/usr/local/lib/python3.6/dist-packages/_pytest/_io/saferepr.py", line 36 in repr 
Jan 29 00:36:45   File "/usr/local/lib/python3.6/dist-packages/_pytest/_io/saferepr.py", line 67 in saferepr 
Jan 29 00:36:45   File "/usr/local/lib/python3.6/dist-packages/_pytest/_code/code.py", line 655 in repr_args 
Jan 29 00:36:45   File "/usr/local/lib/python3.6/dist-packages/_pytest/_code/code.py", line 736 in repr_traceback_entry 
Jan 29 00:36:45   File "/usr/local/lib/python3.6/dist-packages/_pytest/_code/code.py", line 777 in repr_traceback 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

jamesr66a commented on aten/src/ATen/native/quantized/cpu/quant_utils.h:
```diff
@@ -28,6 +28,9 @@ inline TensorQuantizationParams ChooseQuantizationParams(
     bool reduce_range = false) {
+  TORCH_CHECK(
+      min <= max,
+      "In ChooseQuantizationParams, min should be less than or equal to max");
```
if `min == max`, wouldn't `scale` be 0?

jianyuh replied:
Check https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/quant_utils.h#L58-L60
```
  // If scale is 0 or too small so its reciprocal is infinity, we arbitrary
  // adjust the scale to 0.1 . We want to avoid scale's reciprocal being	  // adjust the scale to 0.1 . We want to avoid scale's reciprocal being
  // infinity because some of fbgemm code pre-computes scale's reciprocal to do	  // infinity because some of fbgemm code pre-computes scale's reciprocal to do
  // multiplication instead of division in the time critical part of code.	  // multiplication instead of division in the time critical part of code.
  if (scale == 0.0f || isinf(1.0f / scale)) {	  if (scale == 0.0f || isinf(1.0f / scale)) {
    scale = 0.1;	    scale = 0.1;
  }	  }
```

If `min == max`, we first get `scale = 0` and then adjust it to `0.1` before checking `scale > 0`.

jianyuh committed:
Update on "[pytorch][quant] Add assert for min, max, qmin, qmax for ChooseQuantizationParams"


As Title says.

Differential Revision: [D19610810](https://our.internmc.facebook.com/intern/diff/D19610810/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pytorch][quant] Add assert for min, max, qmin, qmax for ChooseQuantizationParams"


As Title says.

Differential Revision: [D19610810](https://our.internmc.facebook.com/intern/diff/D19610810/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pytorch][quant] Add assert for min, max, qmin, qmax for ChooseQuantizationParams"


As Title says.

Differential Revision: [D19610810](https://our.internmc.facebook.com/intern/diff/D19610810/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pytorch][quant] Add assert for min, max, qmin, qmax for ChooseQuantizationParams"


As Title says.

Differential Revision: [D19610810](https://our.internmc.facebook.com/intern/diff/D19610810/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pytorch][quant] Add assert for min, max, qmin, qmax for ChooseQuantizationParams"


As Title says.

Differential Revision: [D19610810](https://our.internmc.facebook.com/intern/diff/D19610810/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pytorch][quant] Add assert for min, max, qmin, qmax for ChooseQuantizationParams"


As Title says.

Differential Revision: [D19610810](https://our.internmc.facebook.com/intern/diff/D19610810/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pytorch][quant] Add assert for min, max, qmin, qmax for ChooseQuantizationParams"


As Title says.

Differential Revision: [D19610810](https://our.internmc.facebook.com/intern/diff/D19610810/)

[ghstack-poisoned]

supriyar commented:
@jianyuh does this fix the error in `test_quantized_rnn`?
